README.md
<p align="center"><a href="https://github.com/crazy-max/docker-inboxen" target="_blank"><img height="128"src="https://raw.githubusercontent.com/crazy-max/docker-inboxen/master/.res/docker-inboxen.jpg"></a></p>

<p align="center">
  <a href="https://hub.docker.com/r/crazymax/inboxen/"><img src="https://img.shields.io/badge/dynamic/json.svg?label=version&query=$.results[1].name&url=https://hub.docker.com/v2/repositories/crazymax/inboxen/tags&style=flat-square" alt="Latest Version"></a>
  <a href="https://travis-ci.com/crazy-max/docker-inboxen"><img src="https://img.shields.io/travis/com/crazy-max/docker-inboxen/master.svg?style=flat-square" alt="Build Status"></a>
  <a href="https://hub.docker.com/r/crazymax/inboxen/"><img src="https://img.shields.io/docker/stars/crazymax/inboxen.svg?style=flat-square" alt="Docker Stars"></a>
  <a href="https://hub.docker.com/r/crazymax/inboxen/"><img src="https://img.shields.io/docker/pulls/crazymax/inboxen.svg?style=flat-square" alt="Docker Pulls"></a>
  <a href="https://quay.io/repository/crazymax/inboxen"><img src="https://quay.io/repository/crazymax/inboxen/status?style=flat-square" alt="Docker Repository on Quay"></a>
  <a href="https://www.codacy.com/app/crazy-max/docker-inboxen"><img src="https://img.shields.io/codacy/grade/6e477437dfdf48f3a7133d7637d92175.svg?style=flat-square" alt="Code Quality"></a>
  <a href="https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CF9YNTZWJCW3L"><img src="https://img.shields.io/badge/donate-paypal-7057ff.svg?style=flat-square" alt="Donate Paypal"></a>
</p>

## About

🐳 [Inboxen](https://inboxen.org/) image based on Alpine Linux.<br />
If you are interested, [check out](https://hub.docker.com/r/crazymax/) my other 🐳 Docker images!

## Features

### Included

* Alpine Linux 3.10
* Inboxen [WSGI daemon](https://github.com/Inboxen/Inboxen/blob/master/inboxen/wsgi.py) served through [uWSGI](https://uwsgi-docs.readthedocs.io) and Nginx
* Nginx serves /static/ from `app/static` content through [Django collectstatic](https://docs.djangoproject.com/en/1.11/ref/contrib/staticfiles/#collectstatic)
* [Salmon](https://salmon-mail.readthedocs.io) mail server available through port `8823`
* [Celery](http://docs.celeryproject.org/en/latest/index.html) served as a distributed task queue for RabbitMQ
* Automatic DB migration

### From docker-compose

* [Traefik](https://github.com/containous/traefik-library-image) as reverse proxy and creation/renewal of Let's Encrypt certificates
* [PostgreSQL](https://github.com/docker-library/postgres) image as database instance
* [RabbitMQ](https://github.com/docker-library/rabbitmq) image for message queue

## Docker

### Environment variables

#### General

* `TZ` : The timezone assigned to the container and Inboxen (default `UTC`)
* `PUID` : Inboxen user id (default `1000`)
* `PGID`: Inboxen group id (default `1000`)
* `DB_TIMEOUT` : Time in seconds after which we stop trying to reach the PostgreSQL server (default `30`)

#### Inboxen

The following environment variables allow to populate the [settings.ini](https://inboxen.readthedocs.io/en/latest/settings.html) for Inboxen.

**General**

* `IB_ADMIN_EMAILS` : See [general / admin_emails](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_ADMIN_NAMES` : See [general / admin_names](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_ALLOWED_HOSTS` : See [general / allowed_hosts](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `*`)
* `IB_DEBUG` : See [general / debug](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `false`)
* `IB_ENABLE_REGISTRATION` :  See [general / enable_registration](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `false`)
* `IB_ENABLE_USER_EDITING` : See [general / enable_user_editing](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `false`)
* `IB_LANGUAGE_CODE` : See [general / language_code](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `en-gb`)
* `IB_LOGIN_ATTEMPT_COOLOFF` : See [general / login_attempt_cooloff](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `10`)
* `IB_LOGIN_ATTEMPT_LIMIT` :  See [general / login_attempt_limit](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `5`)
* `IB_REGISTER_LIMIT_WINDOW` : See [general / register_limit_window](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `1440`)
* `IB_REGISTER_LIMIT_COUNT` : See [general / register_limit_count](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `100`)
* `IB_SECRET_KEY` : See [general / secret_key](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_SERVER_EMAIL` : See [general / server_email](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `docker-inboxen@localhost`)
* `IB_SITE_NAME` : See [general / site_name](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `Docker Inboxen`)
* `IB_SOURCE_LINK` : See [general / source_link](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `https://github.com/Inboxen/Inboxen`)
* `IB_PER_USER_EMAIL_QUOTA` : See [general / per_user_email_quota](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `0`)

**Inbox**

* `IB_INBOX_LENGTH` : See [inbox / inbox_length](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `5`)
* `IB_INBOX_LIMIT_WINDOW` : See [inbox / inbox_limit_window](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `1440`)
* `IB_INBOX_LIMIT_COUNT` : See [inbox / inbox_limit_count](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `100`)

**Tasks**

* `IB_TASKS_BROKER_URL` : See [tasks / broker_url](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `amqp://guest:guest@localhost:5672//`)
* `IB_TASKS_CONCURRENCY` : See [tasks / concurrency](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `3`)
* `IB_LIBERATION_SENDFILE_METHOD` : See [tasks / liberation / sendfile_method](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `simple`)

**Database**

* `IB_DB_NAME` : See [database / name](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `inboxen`)
* `IB_DB_USER` : See [database / user](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_DB_PASSWORD` : See [database / password](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_DB_HOST` : See [database / host](https://inboxen.readthedocs.io/en/latest/settings.html) setting
* `IB_DB_PORT` : See [database / port](https://inboxen.readthedocs.io/en/latest/settings.html) setting

**Cache**

* `IB_CACHE_TIMEOUT` : See [cache / timeout](https://inboxen.readthedocs.io/en/latest/settings.html) setting (default `300`)

### Volumes

* `/data` : Contains cache, media, logs and liberation data

### Ports

* `8080` : HTTP port of Nginx
* `8823` : SMTP port for Salmon mail server

## Use this image

### Docker Compose

Docker compose is the recommended way to run this image. You can use the following [docker compose template](examples/compose/docker-compose.yml), then run the container :

```bash
touch acme.json
chmod 600 acme.json
docker-compose up -d
docker-compose logs -f
```

## Notes

### Add super user

On first launch, you will have to create a super user to handle administration through `/admin` :

```
$ docker-compose exec --user inboxen inboxen sh -c ". env/bin/activate && ./manage.py createsuperuser --username <username> --email <email>"
Password:
Password (again):
Superuser created successfully.
```

> :warning: Substitute your desired username `<username>` and email address `<email>`

## Upgrade

To upgrade to the latest version, pull the newer image and launch the container. Inboxen will upgrade automatically :

```bash
docker-compose pull
docker-compose up -d
```

## How can I help ?

All kinds of contributions are welcome :raised_hands:!<br />
The most basic way to show your support is to star :star2: the project, or to raise issues :speech_balloon:<br />
But we're not gonna lie to each other, I'd rather you buy me a beer or two :beers:!

[![Paypal](.res/paypal.png)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CF9YNTZWJCW3L)

## License

MIT. See `LICENSE` for more details.


CHANGELOG.md
# Changelog

## deploy-20190609-RC1 (2019/07/15)

* Inboxen deploy-20190609
* Alpine Linux 3.10

## deploy-20190110-RC3 (2019/04/28)

* Bind IPv6

## deploy-20190110-RC2 (2019/01/31)

* Alpine Linux 3.9

## deploy-20190110-RC1 (2019/01/10)

* Inboxen deploy-20190110

## master-25bca97-RC1 (2019/01/08)

* Inboxen/Inboxen@25bca97

## master-d6a9a1d-RC1 (2018/12/01)

* Initial version based on Inboxen/Inboxen@d6a9a1d


build.sh
#!/usr/bin/env bash
set -e

PROJECT=inboxen
BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
BUILD_TAG=docker_build
BUILD_WORKINGDIR=${BUILD_WORKINGDIR:-.}
DOCKERFILE=${DOCKERFILE:-Dockerfile}
VCS_REF=${TRAVIS_COMMIT::7}
RUNNING_TIMEOUT=120
RUNNING_LOG_CHECK="uwsgi entered RUNNING state"

PUSH_LATEST=${PUSH_LATEST:-true}
DOCKER_USERNAME=${DOCKER_USERNAME:-crazymax}
DOCKER_LOGIN=${DOCKER_LOGIN:-crazymax}
DOCKER_REPONAME=${DOCKER_REPONAME:-inboxen}
QUAY_USERNAME=${QUAY_USERNAME:-crazymax}
QUAY_LOGIN=${QUAY_LOGIN:-crazymax}
QUAY_REPONAME=${QUAY_REPONAME:-inboxen}

# Check local or travis
BRANCH=${TRAVIS_BRANCH:-local}
if [[ ${TRAVIS_PULL_REQUEST} == "true" ]]; then
  BRANCH=${TRAVIS_PULL_REQUEST_BRANCH}
fi
DOCKER_TAG=${BRANCH:-local}
if [[ "$BRANCH" == "master" ]]; then
  DOCKER_TAG=latest
elif [[ "$BRANCH" == "local" ]]; then
  BUILD_DATE=
  VERSION=local
fi

echo "PROJECT=${PROJECT}"
echo "VERSION=${VERSION}"
echo "BUILD_DATE=${BUILD_DATE}"
echo "BUILD_TAG=${BUILD_TAG}"
echo "BUILD_WORKINGDIR=${BUILD_WORKINGDIR}"
echo "DOCKERFILE=${DOCKERFILE}"
echo "VCS_REF=${VCS_REF}"
echo "PUSH_LATEST=${PUSH_LATEST}"
echo "DOCKER_LOGIN=${DOCKER_LOGIN}"
echo "DOCKER_USERNAME=${DOCKER_USERNAME}"
echo "DOCKER_REPONAME=${DOCKER_REPONAME}"
echo "QUAY_LOGIN=${QUAY_LOGIN}"
echo "QUAY_USERNAME=${QUAY_USERNAME}"
echo "QUAY_REPONAME=${QUAY_REPONAME}"
echo "TRAVIS_BRANCH=${TRAVIS_BRANCH}"
echo "TRAVIS_PULL_REQUEST=${TRAVIS_PULL_REQUEST}"
echo "BRANCH=${BRANCH}"
echo "DOCKER_TAG=${DOCKER_TAG}"
echo

# Build
echo "### Build"
docker build \
  --build-arg BUILD_DATE=${BUILD_DATE} \
  --build-arg VCS_REF=${VCS_REF} \
  --build-arg VERSION=${VERSION} \
  -t ${BUILD_TAG} -f ${DOCKERFILE} ${BUILD_WORKINGDIR}
echo

echo "### Test"
docker rm -f ${PROJECT} ${PROJECT}-rabbitmq ${PROJECT}-db > /dev/null 2>&1 || true
docker network rm ${PROJECT} > /dev/null 2>&1 || true
docker network create -d bridge ${PROJECT}
docker run -d --network=${PROJECT} --name ${PROJECT}-rabbitmq --hostname ${PROJECT}-rabbitmq rabbitmq:3.7
docker run -d --network=${PROJECT} --name ${PROJECT}-db --hostname ${PROJECT}-db \
  -e "POSTGRES_DB=inboxen" \
  -e "POSTGRES_USER=inboxen" \
  -e "POSTGRES_PASSWORD=inboxen" \
  postgres:9.6
docker run -d --network=${PROJECT} --link ${PROJECT}-db --link ${PROJECT}-rabbitmq \
  -e "IB_SECRET_KEY=${PROJECT}" \
  -e "IB_DB_HOST=${PROJECT}-db" \
  -e "IB_DB_NAME=inboxen" \
  -e "IB_DB_USER=inboxen" \
  -e "IB_DB_PASSWORD=inboxen" \
  -e "IB_TASKS_BROKER_URL=amqp://guest:guest@${PROJECT}-rabbitmq:5672//" \
  --name ${PROJECT} ${BUILD_TAG}
echo

echo "### Waiting for ${PROJECT} to be up..."
TIMEOUT=$((SECONDS + RUNNING_TIMEOUT))
while read LOGLINE; do
  echo ${LOGLINE}
  if [[ ${LOGLINE} == *"${RUNNING_LOG_CHECK}"* ]]; then
    echo "Container up!"
    break
  fi
  if [[ $SECONDS -gt ${TIMEOUT} ]]; then
    >&2 echo "ERROR: Failed to run ${PROJECT} container"
    docker rm -f ${PROJECT} ${PROJECT}-rabbitmq ${PROJECT}-db > /dev/null 2>&1 || true
    exit 1
  fi
done < <(docker logs -f ${PROJECT} 2>&1)
echo

CONTAINER_STATUS=$(docker container inspect --format "{{.State.Status}}" ${PROJECT})
if [[ ${CONTAINER_STATUS} != "running" ]]; then
  >&2 echo "ERROR: Container ${PROJECT} returned status '$CONTAINER_STATUS'"
  docker rm -f ${PROJECT} ${PROJECT}-rabbitmq ${PROJECT}-db > /dev/null 2>&1 || true
  exit 1
fi
docker rm -f ${PROJECT} ${PROJECT}-rabbitmq ${PROJECT}-db > /dev/null 2>&1 || true
echo

if [ "${VERSION}" == "local" -o "${TRAVIS_PULL_REQUEST}" == "true" ]; then
  echo "INFO: This is a PR or a local build, skipping push..."
  exit 0
fi
if [[ ! -z ${DOCKER_PASSWORD} ]]; then
  echo "### Push to Docker Hub..."
  echo "$DOCKER_PASSWORD" | docker login --username "$DOCKER_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${VERSION}
  fi
  docker push ${DOCKER_USERNAME}/${DOCKER_REPONAME}
  echo
fi
if [[ ! -z ${QUAY_PASSWORD} ]]; then
  echo "### Push to Quay..."
  echo "$QUAY_PASSWORD" | docker login quay.io --username "$QUAY_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${VERSION}
  fi
  docker push quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}
  echo
fi


entrypoint.sh
#!/bin/sh

function fixperms() {
  for folder in $@; do
    if $(find ${folder} ! -user inboxen -o ! -group inboxen | egrep '.' -q); then
      echo "Fixing permissions in $folder..."
      chown -R inboxen. "${folder}"
    else
      echo "Permissions already fixed in ${folder}"
    fi
  done
}

function runas_inboxen() {
  su - inboxen -s /bin/sh -c "$1"
}

TZ=${TZ:-UTC}
export PUID=${PUID:-1000}
export PGID=${PGID:-1000}

DB_TIMEOUT=${DB_TIMEOUT:-30}

# Timezone
echo "Setting timezone to ${TZ}..."
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} > /etc/timezone

# Change inboxen UID / GID
echo "Checking if inboxen UID / GID has changed..."
if [[ $(id -u inboxen) != ${PUID} ]]; then
  usermod -u ${PUID} inboxen
fi
if [[ $(id -g inboxen) != ${PGID} ]]; then
  groupmod -g ${PGID} inboxen
fi

# Init
echo "Initializing Inboxen files / folders..."
mkdir -p /data/cache /data/liberation /data/media

# Config
# SECRET_KEY=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 50 | head -n 1)
if [[ -z "$IB_SECRET_KEY" -o -z "$IB_DB_HOST" -o -z "$IB_DB_USER" -o -z "$IB_DB_PASSWORD" ]]; then
  >&2 echo "ERROR: IB_SECRET_KEY, IB_DB_HOST, IB_DB_USER and IB_DB_PASSWORD must be defined"
  exit 1
fi
export PGPASSWORD=${IB_DB_PASSWORD}
dbcmd="psql --host=${IB_DB_HOST} --port=${IB_DB_PORT:-5432} --username=${IB_DB_USER} -lqt"

# https://github.com/Inboxen/Inboxen/blob/master/docs/settings.rst
cat > /app/settings.ini <<EOL
[general]
admin_emails = ${IB_ADMIN_EMAILS}
admin_names = ${IB_ADMIN_NAMES}
allowed_hosts = ${IB_ALLOWED_HOSTS:-*}
debug = ${IB_DEBUG:-false}
enable_registration = ${IB_ENABLE_REGISTRATION:-false}
enable_user_editing = ${IB_ENABLE_USER_EDITING:-false}
language_code = ${IB_LANGUAGE_CODE:-en-gb}
login_attempt_cooloff = ${IB_LOGIN_ATTEMPT_COOLOFF:-10}
login_attempt_limit = ${IB_LOGIN_ATTEMPT_LIMIT:-5}
register_limit_window = ${IB_REGISTER_LIMIT_WINDOW:-1440}
register_limit_count = ${IB_REGISTER_LIMIT_COUNT:-100}
static_root = /app/static
media_root = /data/media
secret_key = ${IB_SECRET_KEY}
server_email = ${IB_SERVER_EMAIL:-docker-inboxen@localhost}
site_name = ${IB_SITE_NAME:-Docker Inboxen}
source_link = ${IB_SOURCE_LINK:-https://github.com/Inboxen/Inboxen}
time_zone = ${TZ}
per_user_email_quota = ${IB_PER_USER_EMAIL_QUOTA:-0}

[inbox]
inbox_length = ${IB_INBOX_LENGTH:-5}
inbox_limit_window = ${IB_INBOX_LIMIT_WINDOW:-1440}
inbox_limit_count = ${IB_INBOX_LIMIT_COUNT:-100}

[tasks]
broker_url = ${IB_TASKS_BROKER_URL:-amqp://guest:guest@localhost:5672//}
concurrency = ${IB_TASKS_CONCURRENCY:-3}

[[liberation]]
path = /data/liberation
sendfile_method = ${IB_LIBERATION_SENDFILE_METHOD:-simple}

[database]
name = ${IB_DB_NAME:-inboxen}
user = ${IB_DB_USER}
password = ${IB_DB_PASSWORD}
host = ${IB_DB_HOST}
port = ${IB_DB_PORT:-5432}

[cache]
backend = file
location = /data/cache
timeout = ${IB_CACHE_TIMEOUT:-300}
EOL
chown inboxen. /app/settings.ini

# Fix perms
echo "Fixing permissions..."
fixperms /app /data/cache /data/liberation /data/logs /data/media

echo "Waiting ${DB_TIMEOUT}s for database to be ready..."
counter=1
while ${dbcmd} | cut -d \| -f 1 | grep -qw ${IB_DB_NAME:-inboxen} > /dev/null 2>&1; [[ $? -ne 0 ]]; do
  sleep 1
  counter=$((counter + 1))
  if [[ ${counter} -gt ${DB_TIMEOUT} ]]; then
    >&2 echo "ERROR: Failed to connect to database on $IB_DB_HOST"
    exit 1
  fi;
done
echo "Database ready!"

unset IB_SECRET_KEY
unset IB_DB_USER
unset IB_DB_PASSWORD
unset IB_DB_HOST
unset PGPASSWORD

# Migrate DB
runas_inboxen ". env/bin/activate && ./manage.py migrate"

exec "$@"
